### PR TITLE
Revise Overview for CuPy v10

### PR DESCRIPTION
**Pull request by kmaehashi**

### CODE REVIEW

File: docs/source/overview.rst [hunk 1]
```diff
@@ -3,56 +3,66 @@
...
+CuPy can run in multi-GPU or cluster environments. The distributed communication package (:mod:`cupyx.distributed`) provides collective and peer-to-peer primitives for ``ndarray``, backed by NCCL.
+
+For users who need more fine-grain control for performance, accessing :doc:`low-level CUDA features <user_guide/cuda_api>` are available:
+
+* **Stream and Event**: CUDA stream and per-thread default stream are supported by all APIs
+* **Memory Pool**: Customizable memory allocator with a built-in memory pool
+* **Profiler**: Supports profiling code using CUDA Profiler and NVTX
+* **Host API Binding**: Directly call CUDA libraries, such as NCCL, cuDNN, cuTENSOR, and cuSPARSELt APIs from Python
+
+CuPy implements standard APIs for data exchange and interoperability, such as `DLPack <https://github.com/dmlc/dlpack>`__, `CUDA Array Interface <https://numba.readthedocs.io/en/stable/cuda/cuda_array_interface.html>`__, ``__array_ufunc__`` (`NEP 13 <https://numpy.org/neps/nep-0013-ufunc-overrides.html>`__), ``_array_function_`` (`NEP 18 <https://numpy.org/neps/nep-0018-array-function-protocol.html>`__), and `Array API Standard <https://data-apis.org/array-api/latest/>`__.
```
Comment (leofang) [inline docs/source/overview.rst:56]:
```suggestion
CuPy implements standard APIs for data exchange and interoperability, such as `DLPack <https://github.com/dmlc/dlpack>`__, `CUDA Array Interface <https://numba.readthedocs.io/en/stable/cuda/cuda_array_interface.html>`__, ``__array_ufunc__`` (`NEP 13 <https://numpy.org/neps/nep-0013-ufunc-overrides.html>`__), ``__array_function__`` (`NEP 18 <https://numpy.org/neps/nep-0018-array-function-protocol.html>`__), and `Array API Standard <https://data-apis.org/array-api/latest/>`__.
```

File: docs/source/user_guide/performance.rst [hunk 2]
```diff
@@ -44,6 +44,26 @@ synchronizing over the end event (see :ref:`cuda_stream_event` for detail). Belo
 
 Additionally, :func:`cupyx.profiler.benchmark` runs a few warm-up runs to reduce timing fluctuation and exclude the overhead in first invocations.
 
+One-Time Overheads
```
Comment (leofang) [inline docs/source/user_guide/performance.rst:47]: Just a nitpick: it seems in this file sections are separated by two empty lines.
```suggestion

One-Time Overheads
```

File: docs/source/user_guide/performance.rst [hunk 3]
```diff
@@ -44,6 +44,26 @@ synchronizing over the end event (see :ref:`cuda_stream_event` for detail). Belo
...
+
+Kernel Compilation
+..................
+
+CuPy uses on-the-fly kernel synthesis; when a kernel call is required, it compiles a kernel code optimized for the shapes and dtypes of given arguments, sends it to the GPU device, and executes the kernel.
+
+CuPy caches the kernel code sent to GPU device within the process, which reduces the kernel transfer time on further calls.
+
+The compiled code is also cached to ``${HOME}/.cupy/kernel_cache`` directory (the path can be overwritten by setting the ``CUPY_CACHE_DIR`` environment variable).
+This allows reusing the compiled kernel binary across the process.
```
Comment (leofang) [inline docs/source/user_guide/performance.rst:66]: ditto
```suggestion
This allows reusing the compiled kernel binary across the process.

```

File: docs/source/user_guide/performance.rst [hunk 4]
```diff
@@ -44,6 +44,26 @@ synchronizing over the end event (see :ref:`cuda_stream_event` for detail). Belo
...
+Context Initialization
+......................
+
+It may take several seconds when calling a CuPy function for the first time in a process.
+This is because CUDA driver creates a CUDA context during the first CUDA API call in CUDA applications.
+
+Kernel Compilation
+..................
+
+CuPy uses on-the-fly kernel synthesis; when a kernel call is required, it compiles a kernel code optimized for the shapes and dtypes of given arguments, sends it to the GPU device, and executes the kernel.
```
Comment (leofang) [inline docs/source/user_guide/performance.rst:61]: I think it's dimensions, not shapes?
```suggestion
CuPy uses on-the-fly kernel synthesis. When a kernel call is required, it compiles a kernel code optimized for the dimensions and dtypes of the given arguments, sends them to the GPU device, and executes the kernel.
```

File: docs/source/user_guide/performance.rst [hunk 5]
```diff
@@ -44,6 +44,26 @@ synchronizing over the end event (see :ref:`cuda_stream_event` for detail). Belo
...
+
+CuPy caches the kernel code sent to GPU device within the process, which reduces the kernel transfer time on further calls.
```
Comment (leofang) [inline docs/source/user_guide/performance.rst:63]:
```suggestion
CuPy caches the kernel code sent to GPU device within the process, which reduces the kernel compilation time on further calls.
```

File: docs/source/user_guide/performance.rst [hunk 6]
```diff
@@ -44,6 +44,26 @@ synchronizing over the end event (see :ref:`cuda_stream_event` for detail). Belo
...
+
+The compiled code is also cached to ``${HOME}/.cupy/kernel_cache`` directory (the path can be overwritten by setting the ``CUPY_CACHE_DIR`` environment variable).
```
Comment (leofang) [inline docs/source/user_guide/performance.rst:65]:
```suggestion
The compiled code is also cached in the directory ``${HOME}/.cupy/kernel_cache`` (the path can be overwritten by setting the :envvar:`CUPY_CACHE_DIR` environment variable).
```

File: docs/source/overview.rst [hunk 7]
```diff
@@ -3,56 +3,66 @@
...
+  * :doc:`Fast Fourier Transform <reference/fft>` (``cupy.fft.*``)
+  * :doc:`Random Number Generator <reference/random>` (``numpy.random.*``)
+
+* **SciPy Routines**
+
+  * :doc:`Discrete Fourier Transforms <reference/scipy_fft>` (``cupyx.scipy.fft.*`` and ``cupyx.scipy.fftpack.*``)
+  * :doc:`Advanced Linear Algebra <reference/scipy_linalg>` (``cupyx.scipy.linalg.*``)
+  * :doc:`Multidimensional Image Processing <reference/scipy_ndimage>` (``cupyx.scipy.ndimage.*``)
+  * :doc:`Sparse Matrices <reference/scipy_sparse>` (``cupyx.scipy.sparse.*``)
+  * :doc:`Sparse Linear Algebra <reference/scipy_linalg>` (``cupyx.scipy.sparse.linalg.*``)
```
Comment (kmaehashi) [inline docs/source/overview.rst:33]:
```suggestion
  * :doc:`Sparse Linear Algebra <reference/scipy_sparse_linalg>` (``cupyx.scipy.sparse.linalg.*``)
```

**Comment by kmaehashi** [comment]
@leofang Thanks, applied suggestions!

File: docs/source/overview.rst [hunk 8]
```diff
@@ -3,56 +3,66 @@
...
+* **Sparse matrices**: :mod:`cupyx.scipy.sparse`
+
+  * 2-D sparse matrix: ``csr_matrix``, ``coo_matrix``, ``csc_matrix``, and ``dia_matrix``
+
+* **NumPy Routines**
+
+  * :doc:`Module-level Functions <reference/routines>` (``cupy.*``)
+  * :doc:`Linear Algebra Functions <reference/linalg>` (``cupy.linalg.*``)
+  * :doc:`Fast Fourier Transform <reference/fft>` (``cupy.fft.*``)
+  * :doc:`Random Number Generator <reference/random>` (``numpy.random.*``)
```
Comment (leofang) [inline docs/source/overview.rst:25]:
```suggestion
  * :doc:`Random Number Generator <reference/random>` (``cupy.random.*``)
```

File: docs/source/overview.rst [hunk 9]
```diff
@@ -3,56 +3,66 @@
...
+CuPy acts as a drop-in replacement to run existing NumPy/SciPy code on `NVIDIA CUDA <https://developer.nvidia.com/cuda-toolkit>`__ or `AMD ROCm <https://www.amd.com/en/graphics/servers-solutions-rocm>`__ platforms.
+
+CuPy provides a ``ndarray``, sparse matrices, and the associated routines for GPU devices, all having the same API as NumPy and SciPy:
+
+* **N-dimensional array**: :class:`cupy.ndarray`
+
+  * Data types (dtypes): boolean (``bool_``), integer (``int8``, ``int16``, ``int32``, ``int64``, ``uint8``, ``uint16``, ``uint32``, ``uint64``), float (``float16``, ``float32``, ``float64``), and complex (``complex64``, ``complex128``)
+  * Supports the semantics identical to :class:`numpy.ndarray`, including basic / advanced indexing and broadcasting
+
+* **Sparse matrices**: :mod:`cupyx.scipy.sparse`
```
Comment (leofang) [inline docs/source/overview.rst:16]: Maybe we should hyperlink to the base class to be consistent in style?
```suggestion
* **Sparse matrices**: :class:`cupyx.scipy.sparse.spmatrix`
```

File: docs/source/overview.rst [hunk 10]
```diff
@@ -3,56 +3,66 @@
...
-make things slower at the first kernel call, though this slow down will be
-resolved at the second execution. CuPy also caches the kernel code sent to GPU
-device within the process, which reduces the kernel transfer time on further
-calls.
+`CuPy <https://github.com/cupy/cupy>`__ is a NumPy/SciPy-compatible array library for GPU-accelerated computing with Python.
+CuPy acts as a drop-in replacement to run existing NumPy/SciPy code on `NVIDIA CUDA <https://developer.nvidia.com/cuda-toolkit>`__ or `AMD ROCm <https://www.amd.com/en/graphics/servers-solutions-rocm>`__ platforms.
+
+CuPy provides a ``ndarray``, sparse matrices, and the associated routines for GPU devices, all having the same API as NumPy and SciPy:
+
+* **N-dimensional array**: :class:`cupy.ndarray`
```
Comment (leofang) [inline docs/source/overview.rst:11]:
```suggestion
* **N-dimensional array** (``ndarray``): :class:`cupy.ndarray`
```

**Review by leofang — COMMENTED** [review]
Thanks, caught a few nits but I think it's in good shape.

**Comment by leofang** [comment]
This should also be changed now that the Overview is rewritten, maybe just redirecting it to the One-Time Overheads section?
截圖 2021-12-06 下午3 06 54

File: docs/source/overview.rst (hunk 9 above)
Comment (kmaehashi) [reply docs/source/overview.rst]: I rather changed `cupy.ndarray` link to `reference/ndarray` instead, as docs for `ndarray` and `spmatrix` are not very interesting 😅

**Review by leofang — APPROVED** [review]
LGTM!

**Review by emcastillo — APPROVED** [review]
https://cupy--6209.org.readthedocs.build/en/6209/overview.html

Looks great! thanks!

**Comment by emcastillo** [comment]
Jenkins, test this please

**Pull request merged**